### PR TITLE
Implement Operation Pool

### PR DESCRIPTION
**Pull request by michaelsproul**
## Issue Addressed

#281

## Proposed Changes

Introduce an `OperationPool` for tracking operations like transfers, slashings, etc and their inclusion in blocks.

## Additional Info

Incomplete. Still needs:

* ~~Support for Attestations~~ [done]
* ~~Validation of Transfers upon insertion (made slightly trickier by not being able to use `verify_transfer` directly, but that's OK)~~ [done]

### CODE REVIEW

**Comment by michaelsproul** [comment]
Current status: mostly complete with a few outstanding issues, namely:

- [ ] We don't yet properly aggregate the _signatures_ of the attestations, as this requires some support in our BLS wrapper (as mentioned [here](https://github.com/sigp/lighthouse/issues/281#issuecomment-474153398))
- [ ] The code for ProposerSlashings (asssuming it's OK) needs to be duped/refactored to support AttesterSlashings as well. Furthermore, we need a mechanism for ensuring proposer slashings and attester slashings returned for a block don't conflict (they can't be produced independently)
- [ ] `get_attestations` needs to also return attestations from the previous epoch (I'm fairly sure, haven't had time to do this yet)
- [ ] Not very well tested... there are tests for deposit handling, but not for anything else. The need to provide valid signed objects to every method is a bit daunting, but with a few more days of effort I think I could do it.
- [ ] Not integrated with the beacon node yet

Other than that there are a few architectural choices that I'm not 100% sure of and could do with some feedback on. In the transaction pool generally it's often useful to be able to validate some transaction _except for checks that might fail now, but succeed in the future_. To deal with this, I've refactored the transaction verification functions so that they have a `verify_X_time_independent_only` variant, which will only check properties that can't become true later. "Time independence" isn't an entirely satisfactory choice of name, but I also can't think of anything better ("future compatible" is the main alternative I've thought of).

File: eth2/operation_pool/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,952 @@
...
+use types::chain_spec::Domain;
+use types::{
+    Attestation, AttestationData, AttesterSlashing, BeaconState, ChainSpec, Deposit, Epoch,
+    ProposerSlashing, Transfer, Validator, VoluntaryExit,
+};
+
+#[cfg(test)]
+const VERIFY_DEPOSIT_PROOFS: bool = false;
+#[cfg(not(test))]
+const VERIFY_DEPOSIT_PROOFS: bool = false; // TODO: enable this
```
Comment (michaelsproul) [inline eth2/operation_pool/src/lib.rs]: @paulhauner Why was this disabled? Safe to re-enable?

Comment (paulhauner) [reply eth2/operation_pool/src/lib.rs]: For records sake, we'll leave this disabled until we add deposit proof generation to `test_harness`

**Comment by paulhauner** [comment]
Merging this into `testnet-client` so it may later be merged into `master`

**Pull request merged**